socs/db/suprsync.py

```diff
@@ -0,0 +1,355 @@
+        time.sleep(1)
+
+
+        if res.returncode != 0:
+            self.log.error("Rsync failed for file {path}!\nstderr: {err}",
+                           path=file.local_path, err=res.stderr.decode())
+            return False
+
+        res = self.run_on_remote(['md5sum', remote_path])
+        if res.returncode != 0:
+            return False
+        else:
+            return res.stdout.decode().split()[0]
+
+    def handle_file(self, file, session):
+        """
+        Handles operation of an un-removed SupRsyncFile. Will attempt to copy,
+        calculate the remote md5sum, and remove from the local host if enough
+        time has passed.
+
+        Args
+        ----
+        file : SupRsyncFile
+            file to be copied
+        """
+        # File must have been removed for some reason
+        if not os.path.exists(file.local_path):
+            self.log.warn(
+                "File {path} does not exist! Setting removed time to 0.",
+                path=file.local_path
+            )
+            file.removed = 0
+            return
+
+        # Check that file md5sum matches whatever it was before
+        current_md5sum = get_md5sum(file.local_path)
+        if current_md5sum != file.local_md5sum:
+            self.log.warn(
+                "Md5sum of {path} has changed! Resetting it in the db and "
+                "trying to copy", path=file.local_path)
+            file.local_md5sum = current_md5sum
+
+        if file.local_md5sum != file.remote_md5sum:
+            self.log.info("Copying file: {path}", path=file.local_path)
+            md5sum = self.copy_file(file)
+            if md5sum is False:  # Copy command failed with exit code != 0
+                self.log.warn(
+                    "Failed to copy {local} (attempts: {c})",
+                    local=file.local_path, c=file.failed_copy_attempts
+                )
+                file.failed_copy_attempts += 1
+                return
+
+            file.remote_md5sum = md5sum
+            file.copied = time.time()
+            if md5sum != file.local_md5sum:
+                self.log.warn(
+                    "Copied {local} but md5sum does not match (attempts: {c})",
+                    local=file.local_path, c=file.failed_copy_attempts
+                )
+                file.failed_copy_attempts += 1
+            else:
+                dest = os.path.join(self.remote_basedir, file.remote_path)
+                if self.ssh_host is not None:
+                    dest = self.ssh_host + ':' + dest
+                self.log.info("Successfully copied {local} to {dest}",
+                               local=file.local_path, dest=dest)
+
+        if self.delete_after is None:
+            return
+
+        if file.local_md5sum == file.remote_md5sum:
+            if time.time() - file.timestamp > self.delete_after:
+                self.log.info(f"Deleting {file.local_path}")
+                os.remove(file.local_path)
```

Comment:
I can't think of a way around this if the docker doesn't have permissions to operate on the file, except that if the agent doesn't have file permissions the delete_after param probably shouldn't be set. I can make a note of this in the docs though, unless you have a better idea of how this should be handled.

Reply:
Right, there isn't much to do to solve it except fix the permissions. What we can do here though is catch the `PermissionError` and log that we couldn't delete the file so that a one off permission error doesn't crash the `run` Process.

Reply:
Works for me